### ISSUE TITLE
@CallSuper - overriding method implementations automatically invoke super.method()

### ISSUE THREAD

**Issue by DaBlick** [rzwitserloot/lombok#1031]
This is a proposal.    I will try to implement myself if I have time but I know I couldn't start on this for at least several months due to work commitments.

It is very common in Java to write methods in a base class that "should" be called by all overriding implementations of the method, in the same way the superclass constructors are invoked.

I would like to see a Lombok annotation such that when present on a non-private method, a call to the base-class implementation is inserted by Lombok in all overriding implementations, similar to what we get automatically from the compiler with constructors.

I recognize there would be limitations to this.  It would only impact subclasses compiled as part of the project (and not those imported from Jars).    

Another possibility is to have a mode where it issues a warning - there might be an associated @Ignore as well.

**Comment by Xfel** [comment]
The problem here is: At which point should it insert the super call? At the start of the method? At the end of the method?

**Comment by DaBlick** [comment]
I think the answer there is clear:  at the start of the method.

Why?

1. It is consistent with what constructors do

2. "the end of the method" is not well defined.

What if there are multiple return statements?   And even if you were to define "end" as the last source line of the method, I'm not sure that the implementation can know (or be guaranteed) what point in the bytecode corresponds to "the last source line".     I think javac is free to reorganize the flow so long as semantics are preserved.

At the very least, the first implementation of this could be "the start of the method".

**Comment by phcouton** [comment]
I'm a little suspicious for methods that return values. If `@CallSuper` generates something like `return super.foo ()` at the start of the method, then overriding the method is useless, isn't it?

**Comment by Maaartinus** [comment]
> "the end of the method" is not well defined.

It is, just use `finally` (e.g.,  `@Cleanup` does it). Bytecode doesn't matter (and Lombok doesn't work with bytecode, it works on the AST).

Doing something at the method end is what the destructors in C++ do.

> At the very least, the first implementation of this could be "the start of the method".

Sure. Calling it at the end could be an option.

> I'm a little suspicious for methods that return values. If @CallSuper generates something like return super.foo () at the start of the method, then overriding the method is useless, isn't it?

Something like `val $super = super.foo ();` would have to be generated instead.

**Comment by DaBlick** [comment]
phcouton.   You raise a good point about return values.  I have a couple of thoughts on this.

First, I think even if `@CallSuper` didn't support methods with return values, it would be valuable.  For that matter, I think it would be valuable even if all it did was issue a compiler diagnostic if the subclass implementation did not call the superclass implementation.   In short, it would have a value/purpose similar to `@Override'.

But let's talk about some possibilities with methods that do return a value.

It is common in this coding pattern for the base class to create something and the subclass to add things to it.  So I think supporting non-void methods is valuable.   

I think Maartinus' proposal to provide a handle like $super is good.   It might be a good option for $super to be the default, but have a parameter to the annotation that allows you to choose something else.

**Comment by DaBlick** [comment]
Maartinus,

I would say I'm a little concerned about using `finally`.    I'm not certain of this, but I think introducing a try block affects performance.  I think it's ok for `@Cleanup` because the right way to ensure you clean up IS a `finally`.    

Also, I think the issue I raised occurs with both AST and bytecode.    I doubt either preserve the source line ordering.    You will have a AST with multiple "returns" -- can you identify from the AST which one is "the" return generated at the last line of the method's source code?    Is that guaranteed?  

Bottom line, is that if this were to be added to Lombok, I would advocate deferring support for putting the super.method() call at returns rather than at the beginning.    I think it has definitional and implementation challenges.  But more importantly, I don't think it's a common use-case and thus has limited value.

**Comment by DaBlick** [comment]
Sorry to bombard with multiple comments... but one more thing.

There is a design choice here.   Remember that the `@SuperClass` annotation is on the super (base) class implementation.    

One approach is to have it automatically generate the calls in subclass methods.  I think it should also NOT generate the call if it finds a user-coded call to the super.method().    That is, it should NOT generate a redundant call.

Another approach is to NOT have it generate any code in subclasses, but rather to issue a diagnostic if the user didn't (manually) code a call.      I think one can make a reasonable argument that affecting NON-annotated code is dangerous.   (That's why, above, I recommend not generating the call if the user manually wrote one.)

Now, if you lean towards the second "approach" here, you can reasonably argument that this is outside the domain of Lombok.     Lombok is for automatic generation of boilerplate code.  Not assertions.  It is really more of an `@override` like annotation that we "wish" Java provided.  However... Lombok provides the best platform on wish to add this and I think it would be a VERY valuable feature.  

It is a very common coding pattern, but the lack of this sort of support makes it a dangerous coding pattern.

**Comment by mpasacrita** [comment]
I agree that this would be a very useful.   Right now, we rely on the developer who implements subclasses to remember to call the superclass method its' overriding.  This would add architectural enforcement.

Although DaBlick is right, that it may be outside the core values of Lombok, but breaking the rules is part of life.  :)   Maybe we can make a subproject that uses Lombok for architects and dev leads to make annotations which enforce coding standards.

**Comment by jbrunel** [comment]
+1 for this proposal. We just came into a case where the subclass and parents methods were updated and we had to update the `super.whatever()` line to fit with the new parameters in every subclass.

For me, this would have the same added value as `@AllArgsConstructor`.

**Comment by kalani96746** [comment]
+1

**Comment by Maaartinus** [comment]
@DaBlick

I guess, try-finally is (mostly) to free, see [this SO answer][1].

The AST stored line numbers for debugging. This is sometimes a pain in Lombok as Lombok uses the line number of the line containing the annotation, which may be elsewhere (clicking on `getX` leads you to `@Getter` on the class, which is logical, but usually not what you want).

> I would advocate deferring support for putting the super.method() call at returns

Sure, that's how I'd proceed.

> Another approach is to NOT have it generate any code in subclasses, but rather to issue a diagnostic

Yes! That's surely simpler and safer. I'd only require that there's a super call somewhere in the method. It needn't be executed, it only must be present. That's IMHO enough as the problem we're solving is forgetting.

> It is a very common coding pattern, but the lack of this sort of support makes it a dangerous coding pattern.

That's why I always avoid it by making the method `final` and calling a new `abstract` method. It's ugly, but fool-proof.

[1]: https://stackoverflow.com/a/141652/581205

**Comment by DaBlick** [comment]
@Maaartinus 

> The AST stored line numbers for debugging. 

Not always (depends on compilation options), but the larger point here is that I assure you (as someone who spent 26 years developing compilers) that "line numbers" are irrelevant in defining "end of the method".     

Without going into a lot of compiler theory here, is that in order to enforce the notion of @CallSuper, you must identify either "a place" that is guaranteed to execute on ALL possible paths through the method, or define "end of the method" as being more than one place.     The former is known in compiler theory as a "post dominator".    A dominator  of a point in the code A is defined as a point in the code that is guaranteed to execute on all paths through the method that include A   The beginning of the method is a **dominator** for ALL points in the method so every such point HAS (at least) one dominator.

A "**post dominator**" for A is a place that is guaranteed to execute AFTER the execution of A on all possible paths through the method that include A.       Because of return statements, there is no guarantee that a post dominator exists.   (At least, unless the compiler inserts merge points for returns that branch to a common point, which is not efficient.)

Bottom line is, I don't really see the "end of the method" thing as useful, and even if we don't agree on the theory suggesting it is hard or impossible, I don't think it's a good argument to NOT do CallSuper with "beginning of method" semantics.   I personally have never had the need for "end of method" semantics arise, but beginning of method semantics are common - even built into the language itself in the form of constructors calling their super constructors.

I appreciate all the inputs people have given.
